src/haz3lweb/Main.re

```diff
@@ -123,13 +123,27 @@ module App = {
     Async_kernel.Deferred.return(state);
   };
 
-  let create = (model: Incr.t(Haz3lweb.Model.t), ~old_model as _, ~inject) => {
+  let create =
+      (
+        model: Incr.t(Haz3lweb.Model.t),
+        ~old_model: Incr.t(Haz3lweb.Model.t),
+        ~inject,
+      ) => {
     open Incr.Let_syntax;
-    let%map model = model;
+    let%map model = model
+    and old_model = old_model;
     Component.create(
       ~apply_action=apply(model),
       model,
       Haz3lweb.Page.view(~inject, ~handlers, model),
+      ~on_display=(_, ~schedule_action as _) => {
```

Comment:
on_display hook runs on every *frame* even when there isn't anything new to paint (so 30+ times per second ideally), so even just doing a diff between old and new zippers in the model is going to cause performance drain. I think we can still use this trigger but we have to only do the work here once per model change, not once per frame.